feat: add SRC-16 Typed Structured Data (#161)
FuelLabs/sway-standards: docs/src/src-16-typed-structured-data.md

This standard aims to:

* Provide a secure, standardized method for hashing structured data
* Enable clear presentation of structured data for user verification during signing
* Support complex data types that mirror Sway structs
* Enable domain separation to prevent cross-protocol replay attacks
* Define a consistent encoding scheme for structured data types
* Remain stateless, not requiring any storage attributes to enable use across all Fuel program types.

## Prior Art

This standard uses ideas from [Ethereum's EIP-712 standard](https://eips.ethereum.org/EIPS/eip-712), adapting its concepts for the Fuel ecosystem. EIP-712 has proven successful in enabling secure structured data signing for applications like the various browser based wallets and signers that are utilized throughout various DeFi protocols.

## Specification

### Definition of Typed Structured Data 𝕊:

The set of structured data 𝕊 consists of all instances of struct types that can be composed from the following types:

Atomic Types:
```sway

> bytes1 to bytes32

Reviewer: What would this look like in practice? Sway does not have a `bytes1` type

Author: Thankyou for pointing out this mistake. This has been corrected to:

Atomic Types:
```sway
u8 to u256
bool
b256 (hash)
```

Dynamic Types:
```sway
Bytes   // Variable-length byte sequences
String  // Variable-length strings
```

Reference Types:

Arrays (both fixed size and dynamic)
Structs (reference to other struct types)

u8 to u256
bool
b256 (hash)
```

Dynamic Types:
```sway
Bytes
String
```

Reference Types:

Arrays (both fixed size and dynamic)
Structs (reference to other struct types)

Example struct definition:

```sway
struct Mail {
    from: b256,
    to: b256,
    contents: String,
}
```

### Domain Separator Encoding

The domain separator provides context for the signing operation, preventing cross-protocol replay attacks. It is computed as hashStruct(domain) where domain is defined as:

```sway
pub struct SRC16Domain {
    name: String,               // The protocol name (e.g., "MyProtocol")
    version: String,            // The protocol version (e.g., "1")
    chain_id: u64,              // The Fuel chain ID

>     verifying_contract: b256,   // The contract address that will verify the signature

Reviewer 2: ```suggestion
    verifying_contract: ContractId,   // The contract id that will verify the signature
```

}
```

The encoding follows this scheme:

* Add SRC16_DOMAIN_TYPE_HASH

> * Add Keccak256 hash of name string

Reviewer 2: Does this have to be keccak? I believe sha256 is cheaper on fuel.

Author: I doesn't need to be. But to better economise the use of a common hash function and not have to write double the amount of encoding methods for `DataEncoder` I think it would be good to use Keccak256. 

I did a quick gas analysis with forc test using SHA256 and Keccak256 for the SRC16Domain and the difference between the two was around 1.36%. See [gas_analysis](https://github.com/CatspersCoffee/src16-research-and-development/blob/main/contracts/tests/gas_analysis/comparison.md)

* Add Keccak256 hash of version string
* Add chain ID as 32-byte big-endian

> * Add verifying contract address as 32 bytes

Reviewer 2: ```suggestion
* Add verifying contract id as 32 bytes
```

Author: Corrected, thankyou.

The set of structured data 𝕊 consists of all instances of struct types that can be composed from the following types:

Atomic Types:
```sway
u8 to u256
bool
b256 (hash)
```

Dynamic Types:
```sway
Bytes   // Variable-length byte sequences
String  // Variable-length strings
```

Reference Types:

Arrays (both fixed size and dynamic)
Structs (reference to other struct types)

Example struct definition:

```sway
struct Mail {
    from: Address,
    to: Address,
    contents: String,
}
```

### Domain Separator Encoding

The domain separator provides context for the signing operation, preventing cross-protocol replay attacks. It is computed as hashStruct(domain) where domain is defined as:

```sway
pub struct SRC16Domain {
    name: String,                   // The protocol name (e.g., "MyProtocol")
    version: String,                // The protocol version (e.g., "1")
    chain_id: u64,                  // The Fuel chain ID
    verifying_contract: ContractId, // The contract id that will verify the signature
}
```

The encoding follows this scheme:

* Add SRC16_DOMAIN_TYPE_HASH
* Add Keccak256 hash of name string
* Add Keccak256 hash of version string

> * Add chain ID as 32-byte big-endian

Reviewer 2: This mentions 32 bytes, but the above SRC16Domain lists it at a u64. Should this be corrected to u256?

Author: Corrected to add more information around the padding of the `u64` value into the `u256` type on Fuel.

* Add verifying contract id as 32 bytes

## Type Encoding

Each struct type is encoded as name ‖ "(" ‖ member₁ ‖ "," ‖ member₂ ‖ "," ‖ … ‖ memberₙ ")" where each member is written as type ‖ " " ‖ name.

Example:

```
Mail(address from,address to,string contents)
```

## Data Encoding

### Definition of hashStruct

The hashStruct function is defined as:

hashStruct(s : 𝕊) = keccak256(typeHash ‖ encodeData(s))
where:

> * typeHash = keccak256(encodeType(typeOf(s)))

Reviewer 2: ```suggestion
* type_hash = keccak256(encode_type(type_of(s)))
```

* ‖ represents byte concatenation

> * encodeType and encodeData are defined below

Reviewer 2: ```suggestion
* encode_type and encode_data are defined below
```

> ### Definition of encodeData

Reviewer 2: ```suggestion
### Definition of encode_data
```

The encoding of a struct instance is enc(value₁) ‖ enc(value₂) ‖ … ‖ enc(valueₙ), the concatenation of the encoded member values in the order they appear in the type. Each encoded member value is exactly 32 bytes long.

The values are encoded as follows:

Atomic Values:
* Boolean false and true are encoded as u64 values 0 and 1, padded to 32 bytes
* Addresses, ContractId, Identity, and b256 are encoded directly as 32 bytes
* Unsigned Integer values (u8 to u256) are encoded as big-endian bytes, padded to 32 bytes

Dynamic Types:
* Bytes and String are encoded as their Keccak256 hash

Reference Types:
* Arrays (both fixed and dynamic) are encoded as the Keccak256 hash of their concatenated encodings
* Struct values are encoded recursively as hashStruct(value)

The implementation of `TypedDataHash` for `𝕊` SHALL utilize the `DataEncoder` for encoding each element of the struct based on its type.

## Final Message Encoding

The encoding of structured data follows this pattern:

> encode(domainSeparator : 𝔹²⁵⁶, message : 𝕊) = "\x19\x01" ‖ domainSeparator ‖ hashStruct(message)

Reviewer 2: ```suggestion
encode(domain_separator : 𝔹²⁵⁶, message : 𝕊) = "\x19\x01" ‖ domain_separator ‖ hash_struct(message)
```

where:

* \x19\x01 is a constant prefix
* ‖ represents byte concatenation

> * domainSeparator is the 32-byte hash of the domain parameters

Reviewer 2: ```suggestion
* domain_separator is the 32-byte hash of the domain parameters
```

> * hashStruct(message) is the 32-byte hash of the structured data

Reviewer 2: ```suggestion
* hash_struct(message) is the 32-byte hash of the structured data
```

#### Fixed Bytes
- EIP712's `bytes32` maps directly to Sway's `b256`
- Encoded using `encode_b256` in the DataEncoder
- Both standards handle 32-byte values identically
- Smaller fixed byte arrays (bytes1 to bytes31) are not supported in SRC16

#### Address Types
- EIP712 uses 20-byte Ethereum addresses
- When encoding an EIP712 address, SRC16:
  - Takes only rightmost 20 bytes from a 32-byte Fuel Address
  - Pads with zeros on the left for EIP712 compatibility
  - Example: Fuel Address of 32 bytes becomes rightmost 20 bytes in EIP712 encoding

#### ContractId Handling
- ContractId is unique to Fuel/SRC16 (no equivalent in EIP712)
- When encoding for EIP712 compatibility:
  - Uses rightmost 20 bytes of ContractId
  - Particularly important in domain separators where EIP712 expects a 20-byte address

#### Domain Separator Compatibility
```rust
// SRC16 Domain (Fuel native)
pub struct SRC16Domain {
    name: String,                   // Same as EIP712
    version: String,                // Same as EIP712
    chain_id: u64,                  // Fuel chain ID
    verifying_contract: ContractId, // Full 32-byte ContractId
}

// EIP712 Domain (Ethereum compatible)

> pub struct EIP712Domain {

Reviewer 2: Is this not missing a salt field?

Author: Corect, the `salt` field is optional (as are all of the fields) in EIP712. I have chosen to make all other fields mandatory when using `EIP712Domain` on Fuel and omit `salt`.
